sqlmesh/core/model/cache.py

```diff
@@ -133,47 +138,65 @@ def _entry_name(model: SqlModel) -> str:
         return f"{model.name}_{crc32(hash_data)}"
 
 
-def optimized_query_cache_pool(optimized_query_cache: OptimizedQueryCache) -> ProcessPoolExecutor:
+def optimized_query_cache_pool(*args: t.Any) -> ProcessPoolExecutor:
     return ProcessPoolExecutor(
         mp_context=mp.get_context("fork"),
         initializer=_init_optimized_query_cache,
-        initargs=(optimized_query_cache,),
+        initargs=args,
         max_workers=c.MAX_FORK_WORKERS,
     )
 
 
-@t.overload
-def load_optimized_query_cache(
-    model_or_tuple: t.Tuple[Model, T],
-) -> t.Tuple[T, t.Optional[str]]: ...
+_optimized_query_cache: t.Optional[OptimizedQueryCache] = None
+_audits: t.Optional[t.Dict[str, ModelAudit]] = None
 
 
-@t.overload
-def load_optimized_query_cache(model_or_tuple: Model) -> t.Tuple[str, t.Optional[str]]: ...
+def _init_optimized_query_cache(
+    optimized_query_cache: t.Optional[OptimizedQueryCache] = None,
+    audits: t.Optional[t.Dict[str, ModelAudit]] = None,
+) -> None:
+    global _optimized_query_cache, _nodes, _audits
+    _optimized_query_cache = optimized_query_cache
+    _audits = audits
 
 
-def load_optimized_query_cache(model_or_tuple):  # type: ignore
+def load_optimized_query(
+    model_snapshot_id: t.Tuple[Model, SnapshotId],
+) -> t.Tuple[SnapshotId, t.Optional[str]]:
     assert _optimized_query_cache
-
-    if isinstance(model_or_tuple, _Model):
-        model = model_or_tuple
-        key = None
-    else:
-        model, key = model_or_tuple
+    model, snapshot_id = model_snapshot_id
 
     if isinstance(model, SqlModel):
         entry_name = _optimized_query_cache.put(model)
     else:
         entry_name = None
-    return key or model.fqn, entry_name
+    return snapshot_id, entry_name
 
 
-_optimized_query_cache: t.Optional[OptimizedQueryCache] = None
+def load_optimized_query_and_mapping(
+    model: Model, mapping: t.Dict
+) -> t.Tuple[str, t.Optional[str], str, str, t.Dict]:
+    assert _optimized_query_cache
+    assert _audits is not None
 
+    schema = MappingSchema(normalize=False)
+    for parent, columns_to_types in mapping.items():
+        schema.add_table(parent, columns_to_types, dialect=model.dialect)
+    model.update_schema(schema)
```

Comment:
I'm not sure I follow: what's the point of doing the update schema for individual model on the executor side?

Reply:
update_schema is an expensive operation it's 

Models X Upstreams per Model X Columns Per upstream

Reply:
it's done on a different data type, so it's free, the expensive work is done in cores, then we can just set it in teh main process